cpp/src/stream_compaction/stream_compaction.cu

```diff
@@ -108,4 +115,118 @@ gdf_column drop_nulls(gdf_column const &input) {
   }
 }
 
+
+namespace detail {
+struct tuple_reduced_and
+{
+    __host__ __device__ bool operator()(const thrust::tuple<bool,bool> &x) const {
+        return thrust::get<0>(x) && thrust::get<1>(x); }
+}; 
+
+template<typename DerivedPolicy,
+         typename InputIterator,
+         typename OutputIterator,
+         typename BinaryPredicate>
+__host__ __device__
+  OutputIterator non_duplicates_copy(thrust::execution_policy<DerivedPolicy> &exec,
+                             InputIterator first,
+                             InputIterator last,
+                             OutputIterator output,
+                             BinaryPredicate binary_pred)
+{
+  thrust::detail::head_flags<InputIterator, BinaryPredicate> stencil1(first, last, binary_pred);
```

Comment:
I'm not familiar with what the `head_flags` and `tail_flags` constructs do. And I'm in general uneasy about using anything from the `detail` namespace of another library. At the very least, these need to be documented on what they do.

Reply:
`head_flags` is used by `thrust::unique()` to identify non-duplicate entries (returns 1 for unique and 1st duplicate entry.)
`tail_flags` returns 1 for unique and last duplicate entry. (thrust also does not have the documentation)
I will add the documentation in our usage.

I used `thrust::detail` namespace to avoid code duplication. I can create the combined functionality here itself. would you prefer that?

Reply:
You shouldn't need to rely on head or tail flags...  Those are an internal implementation detail of Thrust. 

If you want to know if an element of a sorted array is a duplicate, just compare it to the element before and the element after.  Also, how do you handle null bit masks with `thrust::copy_if`? Don't you need to use `cudf::copy_if`?

Reply:
I replaced it with lamda functions and `thrust::copy_if`.
`thurst::copy_if` is used to copy only indices, not columns. For copying columns, I use `cudf::gather`.